substrate/frame/broker/src/migration.rs

```diff
@@ -156,6 +162,244 @@ mod v3 {
+		pub contribution_timeout: Timeslice,
+	}
+
+	#[storage_alias]
+	pub type SaleInfo<T: Config> = StorageValue<Pallet<T>, SaleInfoRecordOf<T>, OptionQuery>;
+	pub type SaleInfoRecordOf<T> =
+		SaleInfoRecord<BalanceOf<T>, frame_system::pallet_prelude::BlockNumberFor<T>>;
+
+	/// The status of a Bulk Coretime Sale.
+	#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+	pub struct SaleInfoRecord<Balance, BlockNumber> {
+		/// The relay block number at which the sale will/did start.
+		pub sale_start: BlockNumber,
+		/// The length in relay chain blocks of the Leadin Period (where the price is decreasing).
+		pub leadin_length: BlockNumber,
+		/// The price of Bulk Coretime after the Leadin Period.
+		pub price: Balance,
+		/// The first timeslice of the Regions which are being sold in this sale.
+		pub region_begin: Timeslice,
+		/// The timeslice on which the Regions which are being sold in the sale terminate. (i.e.
+		/// One after the last timeslice which the Regions control.)
+		pub region_end: Timeslice,
+		/// The number of cores we want to sell, ideally. Selling this amount would result in no
+		/// change to the price for the next sale.
+		pub ideal_cores_sold: CoreIndex,
+		/// Number of cores which are/have been offered for sale.
+		pub cores_offered: CoreIndex,
+		/// The index of the first core which is for sale. Core of Regions which are sold have
+		/// incrementing indices from this.
+		pub first_core: CoreIndex,
+		/// The latest price at which Bulk Coretime was purchased until surpassing the ideal number
+		/// of cores were sold.
+		pub sellout_price: Option<Balance>,
+		/// Number of cores which have been sold; never more than cores_offered.
+		pub cores_sold: CoreIndex,
+	}
+}
+
+pub mod v4 {
+	use super::*;
+
+	type BlockNumberFor<T> = frame_system::pallet_prelude::BlockNumberFor<T>;
+
+	pub trait BlockToRelayHeightConversion<T: Config> {
+		/// Converts absolute value of parachain block number to relay chain block number
+		fn convert_block_number_to_relay_height(
+			block_number: BlockNumberFor<T>,
+		) -> RelayBlockNumberOf<T>;
+
+		/// Converts parachain block length into equivalent relay chain block length
+		fn convert_block_length_to_relay_length(
+			block_number: BlockNumberFor<T>,
+		) -> RelayBlockNumberOf<T>;
+	}
+
+	pub struct MigrateToV4Impl<T, BlockConversion>(PhantomData<T>, PhantomData<BlockConversion>);
+	impl<T: Config, BlockConversion: BlockToRelayHeightConversion<T>> UncheckedOnRuntimeUpgrade
+		for MigrateToV4Impl<T, BlockConversion>
+	{
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<Vec<u8>, sp_runtime::TryRuntimeError> {
+			let (interlude_length, configuration_leadin_length) =
+				if let Some(config_record) = v3::Configuration::<T>::get() {
+					(config_record.interlude_length, config_record.leadin_length)
+				} else {
+					((0 as u32).into(), (0 as u32).into())
```

Comment:
Is there any correct scenario in which this `Configuration` does not exist? should we not err, or panic here?

Reply:
Looking deeper into the code it seems that it is indeed possible, if `configure` call on the pallet is not called yet then `Configuration` will be none

Reply:
I see, then what you have done could make sense! 

but it is also worth taking it into account: this pallet is only used in coretime chain, what is the live state of the coretime chain right now?